File: week2/3/D.java
 // https://leetcode.com/problems/search-in-a-binary-search-tree/
//  https://walkccc.me/LeetCode/problems/0700/s

 public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
  public TreeNode searchBST(TreeNode root, int val) {
    if (root == null)
      return null;
    if (root.val == val)
      return root;
    if (root.val > val)
      return searchBST(root.left, val);
    return searchBST(root.right, val);
  }
}

File: week2/3/D.py
# // https://leetcode.com/problems/search-in-a-binary-search-tree/
# https://walkccc.me/LeetCode/problems/0700/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
#   inputs
#   이진 검색 트리(BST)의 root
#   Int number val

#   Output
#   부분 트리를 반환
#   Or null
  def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
    # condition) end of recursion  
    if not root:
        return None

    # preprocess) root node == vale
    if root.val == val:
       return root 

    # val is root's left side
    if root.val > val:
       return self.searchBST(root.left, val);

    # else
    # val is root's right side
    return self.searchBST(root.right, val)